virt_lightning/api.py

```diff
@@ -0,0 +1,442 @@
+        name=host["name"],
+        backing_on=host["distro"],
+        size=host.get("root_disk_size", 15),
+    )
+    domain.add_root_disk(root_disk_path)
+    networks = host.get("networks", [{}])
+    for i, network in enumerate(networks):
+        if "network" not in network:
+            network["network"] = configuration.network_name
+        if i == 0 and not network.get("ipv4"):
+            network["ipv4"] = hv.get_free_ipv4()
+        network["mac"] = hv.reuse_mac_address(
+            network["network"], host["name"], network.get("ipv4")
+        )
+        domain.attach_network(**network)
+    hv.start(domain, metadata_format=host.get("metadata_format", {}))
+    return domain
+
+
+def up(virt_lightning_yaml, configuration, context="default", **kwargs):
+    """
+    Create a list of VM
+    """
+
+    def myDomainEventAgentLifecycleCallback(
+        conn, dom, state, reason, opaque
+    ):  # noqa: N802
+        if state == 1:
+            logger.info("%s %s QEMU agent found", symbols.CUSTOMS.value, dom.name())
+
+    loop = kwargs.get("loop") or asyncio.get_event_loop()
+    _register_aio_virt_impl(loop)
+
+    conn = libvirt.open(configuration.libvirt_uri)
+    hv = vl.LibvirtHypervisor(conn)
+
+    conn.setKeepAlive(5, 3)
+    conn.domainEventRegisterAny(
+        None,
+        libvirt.VIR_DOMAIN_EVENT_ID_AGENT_LIFECYCLE,
+        myDomainEventAgentLifecycleCallback,
+        None,
+    )
+
+    hv.init_network(configuration.network_name, configuration.network_cidr)
+    hv.init_storage_pool(configuration.storage_pool)
+
+    pool = ThreadPoolExecutor(max_workers=10)
+
+    async def deploy():
+        futures = []
+        for host in virt_lightning_yaml:
+            futures.append(
+                loop.run_in_executor(
+                    pool, _start_domain, hv, host, context, configuration
+                )
+            )
+
+        domain_reachable_futures = []
+        for f in futures:
+            await f
+            domain = f.result()
+            if domain:
+                domain_reachable_futures.append(domain.reachable())
+        logger.info("%s ok Waiting...", symbols.HOURGLASS.value)
+
+        await asyncio.gather(*domain_reachable_futures)
+
+    loop.run_until_complete(deploy())
+    logger.info("%s You are all set", symbols.THUMBS_UP.value)
+
+
+def start(configuration, context="default", enable_console=False, **kwargs):
+    """
+    Start a single VM
+    """
+    conn = libvirt.open(configuration.libvirt_uri)
+    hv = vl.LibvirtHypervisor(conn)
+    hv.init_network(configuration.network_name, configuration.network_cidr)
+    hv.init_storage_pool(configuration.storage_pool)
+    host = {
+        k: kwargs[k] for k in ["name", "distro", "memory", "vcpus"] if kwargs.get(k)
+    }
+    domain = _start_domain(hv, host, context, configuration)
+    if not domain:
+        return
+
+    loop = asyncio.get_event_loop()
+
+    if enable_console:
+        import time
+
+        time.sleep(4)
```

Comment:
What is reason to use `sleep()` function at code? Remove sleep if it possible.

Reply:
I think this call is save because we need to wait for the stream creation.